utils/uevent/events.go

```diff
@@ -0,0 +1,58 @@
+package uevent
+
+import (
+	"encoding/json"
+	"fmt"
+	"slices"
+
+	abci "github.com/cometbft/cometbft/abci/types"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/gogoproto/proto"
+	"golang.org/x/exp/maps"
+)
+
+// EmitTypedEvent takes a typed event and emits it
+func EmitTypedEvent(ctx sdk.Context, tev proto.Message) error {
+	event, err := TypedEventToEvent(tev)
+	if err != nil {
+		return err
+	}
+	ctx.EventManager().EmitEvent(event)
+	return nil
+}
+
+// TypedEventToEvent takes typed event and converts to Event object
+func TypedEventToEvent(tev proto.Message) (ev sdk.Event, err error) {
+	evtType := proto.MessageName(tev)
+
+	var evtJSON []byte
+	evtJSON, err = codec.ProtoMarshalJSON(tev, nil)
+	if err != nil {
+		return
+	}
+
+	var attrMap map[string]any
+	if err = json.Unmarshal(evtJSON, &attrMap); err != nil {
+		return
+	}
+
+	// sort the keys to ensure the order is always the same
+	keys := maps.Keys(attrMap)
+	slices.Sort(keys)
```

Comment:
is this for determinism?
I don't _think_ events are expected to be deterministic anyway so maybe it's not a problem?

Also, wouldn't you have to sort any sub maps recursively?

Reply:
This is copied code from the sdk with the change that it doesn't use a `json.RawMessage` type for the attribute value, as in the [original sdk code](https://github.com/cosmos/cosmos-sdk/blob/bdb9232ed819387cceef59fe635858ec985ff816/types/events.go#L96).
The reason for this is that the original sdk `EmitTypedEvent` adds double quotes to the event string attributes, which sucks when you want to query them or subscribe to them.

[Here](https://github.com/cosmos/cosmos-sdk/pull/7762#issuecomment-744605866) is some mention of the issue.

So yeah, I think there are a bunch of nicer ways to emit typed events than what they implemented in the SDK, and those events should be consistent with the legacy ones (emitted by `EmitEvent`)

Reply:
cool
I would suggest including a github link to the original code in the docstring or body of the function
It just helps in future if something needs to be updated or just in general for people to figure out

Reply:
ah you did it
nice

Reply:
Why don't you just take the string and remove the double quotes at head and tail?
By using `map[string]any` will eliminate the json structure of nested structs

<img width="884" alt="Screen Shot 2024-08-22 at 20 22 16" src="https://github.com/user-attachments/assets/13da64d3-3bb8-4acf-8c4d-0ae718b05e5b">

cc @danwt @omritoptix

Reply:
yeah, I did mention that would happen: https://github.com/dymensionxyz/sdk-utils/pull/5#issue-2477408348

Reply:
yeah, sounds like it would likely be better

Reply:
[here](https://github.com/dymensionxyz/sdk-utils/pull/6/files) it is